File: python/examples/native/extract_from_file/extract_from_file_decorator.py
#!/usr/bin/env python3
"""
============================================================================
EXTRACT_FROM_FILE TOOL — DECORATOR PATTERN
============================================================================

PATTERN: @tool Decorator
────────────────────────────────────────────────────────────────────────────
Wraps Matimo tool calls in a class using the @tool("tool-name") decorator.
The decorator intercepts method calls and routes them through Matimo
automatically — the method body is never executed.

Use this pattern when:
  ✅ Building class-based applications or services
  ✅ Encapsulating file/URL text-extraction logic in strongly-typed wrappers
  ✅ Combining multiple tools in a single service layer
  ✅ Object-oriented design

SETUP:
────────────────────────────────────────────────────────────────────────────
  No API key required — extract_from_file is a built-in core tool.

USAGE:
────────────────────────────────────────────────────────────────────────────
  export MATIMO_AUTO_APPROVE=true
  uv run python native/extract_from_file/extract_from_file_decorator.py

============================================================================
"""

import asyncio
from pathlib import Path

from dotenv import load_dotenv

from matimo import Matimo
from matimo.decorators import set_global_matimo_instance, tool

load_dotenv(Path(__file__).parent.parent.parent / ".env")


# ───────────────────────────────────────────────────────────────────────────
# Service class — each method is auto-routed to the matching Matimo tool
# ───────────────────────────────────────────────────────────────────────────

class FileExtractor:
    """High-level file-extraction service using the @tool decorator pattern."""

    @tool("extract_from_file")
    async def extract_local_file(self, file_path: str, format: str = "auto") -> dict:  # noqa: A002
        """
        Decorator auto-calls matimo.execute('extract_from_file', {...}).

        Args:
            file_path: Path to a local PDF/DOCX/TXT/CSV file
            format: 'auto' | 'pdf' | 'docx' | 'txt' | 'csv'

        Returns:
            Extracted text plus format-specific metadata
        """
        ...

    @tool("extract_from_file")
    async def extract_remote_file(self, file_url: str, format: str = "auto") -> dict:  # noqa: A002
        """Extract text from a remote file over HTTP(S)."""
        ...


# ───────────────────────────────────────────────────────────────────────────
# Main
# ───────────────────────────────────────────────────────────────────────────

async def main() -> None:
    print("\n╔════════════════════════════════════════════════════════╗")
    print("║     Extract From File Tool — Decorator Pattern        ║")
    print("║     (@tool decorators for automatic execution)         ║")
    print("╚════════════════════════════════════════════════════════╝\n")

    print("🚀  Initializing Matimo…")
    matimo = await Matimo.init(auto_discover=True)
    set_global_matimo_instance(matimo)
    print("✅  Matimo initialized\n")

    extractor = FileExtractor()
    sample_txt = Path(__file__).parent / "sample-notes.txt"
    sample_txt.write_text(
        "Matimo makes it easy to define tools once in YAML and run them anywhere."
    )

    try:
        print("1. Extracting sample-notes.txt\n")
        result1 = await extractor.extract_local_file(str(sample_txt), format="txt")
        if result1.get("success"):
            metadata = result1.get("metadata", {})
            print(f"Format: {result1.get('format_detected')}")
            print(f"Word count: {metadata.get('word_count')}")
            print(f"Extracted text: {result1.get('extracted_text')}")
        else:
            print(f"Error: {result1.get('error')}")
        print("---\n")

        print("✅  Decorator example completed successfully")
    except Exception as error:
        print(f"❌  Error: {error}\n")
    finally:
        if sample_txt.exists():
            sample_txt.unlink()


if __name__ == "__main__":
    asyncio.run(main())

File: typescript/examples/tools/extract-from-file/extract-from-file-decorator.ts
import {
  MatimoInstance,
  setGlobalMatimoInstance,
  tool,
  getGlobalApprovalHandler,
  type ApprovalRequest,
} from '@matimo/core';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import * as readline from 'readline';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

/**
 * Create an interactive approval callback for file extraction operations.
 */
function createApprovalCallback() {
  return async (request: ApprovalRequest): Promise<boolean> => {
    const isInteractive = process.stdin.isTTY;

    console.info('\n' + '='.repeat(70));
    console.info('🔒 APPROVAL REQUIRED FOR FILE EXTRACTION');
    console.info('='.repeat(70));
    console.info(`\n📋 Tool: ${request.toolName}`);
    console.info(`📝 Description: ${request.description || '(no description provided)'}`);
    if (request.params.filePath) console.info(`   filePath: ${request.params.filePath}`);
    if (request.params.fileUrl) console.info(`   fileUrl: ${request.params.fileUrl}`);

    if (!isInteractive) {
      console.info('\n❌ REJECTED - Non-interactive environment (no terminal)');
      console.info('\n💡 To enable auto-approval in CI/scripts:');
      console.info('   export MATIMO_AUTO_APPROVE=true');
      console.info('\n' + '='.repeat(70) + '\n');
      return false;
    }

    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

    return new Promise((resolve) => {
      console.info('\n❓ User Action Required');
      rl.question('   Type "yes" to approve or "no" to reject: ', (answer) => {
        const approved = answer.toLowerCase() === 'yes' || answer.toLowerCase() === 'y';
        console.info(
          approved ? '   ✅ Operation APPROVED by user' : '   ❌ Operation REJECTED by user'
        );
        console.info('='.repeat(70) + '\n');
        rl.close();
        resolve(approved);
      });
    });
  };
}

/**
 * Example: extract_from_file tool using the @tool decorator pattern.
 */
class FileExtractor {
  @tool('extract_from_file')
  async extractLocalFile(filePath: string, format?: string): Promise<unknown> {
    // Decorator automatically intercepts and executes via Matimo
    return undefined;
  }
}

async function decoratorExample() {
  const matimo = await MatimoInstance.init({ autoDiscover: true });
  setGlobalMatimoInstance(matimo);

  const approvalHandler = getGlobalApprovalHandler();
  approvalHandler.setApprovalCallback(createApprovalCallback());

  console.info('\n' + '='.repeat(70));
  console.info('🚀 Extract From File Tool - Decorator Pattern Example');
  console.info('='.repeat(70));

  const autoApproveEnabled = process.env.MATIMO_AUTO_APPROVE === 'true';
  console.info('\n🔐 APPROVAL CONFIGURATION:');
  console.info(
    autoApproveEnabled
      ? '   ✅ MATIMO_AUTO_APPROVE=true — extraction requests will be AUTO-APPROVED'
      : '   ⚠️  INTERACTIVE MODE ENABLED — you will be prompted to approve extractions'
  );

  const sampleTxt = path.join(__dirname, 'sample-notes.txt');
  fs.writeFileSync(
    sampleTxt,
    'Matimo makes it easy to define tools once in YAML and run them anywhere.'
  );

  const extractor = new FileExtractor();

  try {
    console.info('\n1️⃣  EXTRACTING PLAIN TEXT FILE');
    console.info('-'.repeat(70));
    console.info('Extracting sample-notes.txt\n');
    const result = await extractor.extractLocalFile(sampleTxt, 'txt');
    if (result) {
      console.info('✅ Format:', (result as any).format_detected);
      console.info('📊 Word count:', (result as any).metadata?.word_count);
      console.info('📝 Extracted text:', (result as any).extracted_text);
    }
    console.info('---\n');

    console.info('✅ Decorator example completed successfully');
  } catch (error: any) {
    console.error('❌ Error:', error.message);
  } finally {
    if (fs.existsSync(sampleTxt)) fs.unlinkSync(sampleTxt);
  }
}

decoratorExample();
